internal/auth/interface.go

```diff
@@ -0,0 +1,197 @@
+// limitations under the License.
+
+package auth
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"strings"
+	"sync"
+
+	"github.com/lestrrat-go/jwx/v2/jwt"
+)
+
+// Identity represents a particular user's identity in a particular trust domain
+// (represented by an IdentityProvider).
+type Identity struct {
+	// UserID is a stable unique identifier for the user.  This may be a large
+	// integer or a UUID, rather than something human-readable.
+	//
+	// For KeyCloak, this is `sub`.
+	UserID string
+	// HumanName is a human-readable name.  Because humans are fickle, these may
+	// not be unique or stable over time, though they should be unique at any
+	// particular time.  For example, Alex may change their handle from
+	// "alexsmith" to "alexawesome" after a life change, and someone else might
+	// enroll the "alexsmith" handle.  If you are storing data, you want UserID,
+	// not HumanName.  If you are presenting data, you probably want HumanName.
+	//
+	// For KeyCloak, this is `preferred_username`.
+	HumanName string
+	// Provider is the identity provider that vended this identity.  Note that
+	// UserID and HumanName are only unique within the context of a single
+	// identity provider.
+	Provider IdentityProvider
+}
+
+// String implements strings.Stringer, and also provides a stable storage
+// representation of the Identity.
+func (i *Identity) String() string {
+	if i == nil {
+		return ""
+	}
+	p := i.Provider.String()
+	if p == "" {
+		// Special case for provider registered as "".
+		return i.UserID
+	}
+	return fmt.Sprintf("%s/%s", i.UserID, p)
+
+}
+
+// Human returns a human-readable representation of the identity, suitable for
+// presentation to humans.
+func (i *Identity) Human() string {
+	if i == nil {
+		return "<unknown>"
+	}
+	p := i.Provider.String()
+	if p == "" {
+		// Special case for provider registered as "".
+		return i.UserID
+	}
+	return fmt.Sprintf("%s/%s", i.HumanName, p)
+}
+
+// Resolver is an interface for resolving human-readable or stable identifiers
+// from either JWTs or stored strings
+type Resolver interface {
+
+	// Validate validates a token and returns an underlying identity representation
+	// suitable for use in authz calls.  This _probably_ reads data from the token,
+	// but could fetch from an external provider.
+	Validate(ctx context.Context, token jwt.Token) (*Identity, error)
+
+	// Resolve takes either a human-readable identifier or a stable identifier and
+	// returns the underlying identity.  This may involve looking up or defining
+	// the identity in the remote identity provider.
+	//
+	// For Keycloak + GitHub, this may define a new user in Keycloak based on
+	// GitHub user data if the user is not already known to Keycloak.
+	Resolve(ctx context.Context, id string) (*Identity, error)
+}
+
+// IdentityProvider provides an abstract interface for looking up identities
+// in a remote identity provider.
+type IdentityProvider interface {
```

Comment:
is IdentityProvider assumed to be OIDC-compatible? if so, let's mark that in the docs.

Reply:
If it is, could we use zitadel's `RelyingParty` interface instead of creating our own abstraction?

Reply:
We're already using it in our code base when we create the Keycloak provider

Reply:
I'm not sure that we want to require it.  I imagined this as being decoupled from specific identity provider implementations, so we'd still need something like OIDC (or x.509 client certs) to assert the user's identity.  Once we have their identity, we can then go through this interface to map it to/from database representation.